recipes/nsqadmin.rb

```diff
@@ -13,19 +13,52 @@
 nsq_release = "nsq-#{node['nsq']['version']}-#{node['nsq']['go_version']}"
 
 if node['nsq']['setup_services']
-  template '/etc/init/nsqadmin.conf' do
-    action :create
-    source 'upstart.nsqadmin.conf.erb'
-    mode '0644'
-    # need to stop/start in order to reload config
-    if node['nsq']['reload_services']
-      notifies :stop, 'service[nsqadmin]', :immediately
-      notifies :start, 'service[nsqadmin]', :immediately
+
+  if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04
+
+    template '/etc/systemd/system/nsqadmin.service' do
+      action :create
+      source 'systemd.nsqadmin.conf.erb'
+      mode '0644'
+      notifies :run, 'systemd_unit[nsqadmin.service]', :delayed
+      # need to stop/start in order to reload config
+      if node['nsq']['reload_services']
+        notifies :stop, 'service[nsqadmin]', :immediately
+        notifies :start, 'service[nsqadmin]', :immediately
+      end
+    end
+
+    template "#{node['chef-nsq']['script_dir']}/nsqadmin-start.sh" do
+      action :create
+      source 'nsqadmin-start.sh.erb'
+      mode '0550'
+      owner node['nsq']['nsqadmin']['user']
+      group node['nsq']['nsqadmin']['user']
+    end
+
+    systemd_unit 'nsqadmin.service' do
+      action :nothing
+    end
+
+  else
+    template '/etc/init/nsqadmin.conf' do
+      action :create
+      source 'upstart.nsqadmin.conf.erb'
+      mode '0644'
+      # need to stop/start in order to reload config
+      if node['nsq']['reload_services']
+        notifies :stop, 'service[nsqadmin]', :immediately
+        notifies :start, 'service[nsqadmin]', :immediately
+      end
     end
   end
 
+  provider = Chef::Provider::Service::Upstart
+  if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04
+    provider = Chef::Provider::Service::Systemd
+  end
   service 'nsqadmin' do
-    provider Chef::Provider::Service::Upstart
+    provider provider
```

Comment:
you shouldn't need to specify a provider here at all; chef should pick up the Right One from inspecting pid 1

Reply:
Ahh, I was following the pattern in `nsqlookupd`. Is it safer and better to just remove the `provider` declaration completely?